### PR TITLE
Fix autocomplete behavior in breakpoint widget

### LINKED ISSUE #8825: Suggestions within logpoints not working as expected
### Bug Description:
Autocomplete suggestions within a log message when creating a logpoint have unexpected behavior.
- Suggestions are shown without the interpolated scope ( { } )
- When trying to add interpolation suggestions are only shown after 3 characters
- Trying to autocomplete a found suggestion with enter does add it to the already typed characters, e.g:
 {myVaria}-> auto completes to {myVariamyVariable}
- suggestions sometimes don't make sense in the scope currently working, expected suggestions don't show up

 
### Steps to Reproduce:

1. Open workspace on gitpod.io
2. Add logpoint with type log message
3. Without scope trigger suggestions (ctrl space) results in suggestions to show up
4. Add interpolation via {} trigger suggentions again results in no suggestions

logmessage_suggest

### Additional Information

- Operating System: Win10 (Ms Edge 87.0.664.52)
- Theia Version: Can be reproduced with the gitpod sample applications

### PR DESCRIPTION
**Pull request by kenneth-marut-work**
#### What it does

Fixes #8825 by only enabling autocompletion in Log Points context when brackets are open. Also fixes issues related to caret positioning and suggestion insertion, and updates the interface for provideSuggestionItems which had an incorrect return type leading to all suggestions being broken inside the breakpoints widget (on master), see error below:

image

---

- CQ: [https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23535](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23535)

#### How to test

- Open a source code file of your choice 
- Try inserting a log point inside a function or anywhere you might get autocompletion
- start typing a variable name and notice that there is no autocomplete
- Open a pair of curly brackets and observe that autocomplete works (after typing a single character)
- Accept an autocomplete suggestion with `Enter`, observe that the suggestion is placed exactly where you expect it
-Try adding an expression breakpoint and observe that there are no regressions (autocomplete should work normally)

autocomplete

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/debug/src/browser/editor/debug-breakpoint-widget.tsx [hunk 1]
```diff
@@ -92,23 +93,30 @@ export class DebugBreakpointWidget implements Disposable {
         this.toDispose.push(monaco.modes.CompletionProviderRegistry.register({ scheme: input.uri.scheme }, {
             provideCompletionItems: async (model, position, context, token) => {
                 const suggestions = [];
-                if ((this.context === 'condition' || this.context === 'logMessage')
-                    && input.uri.toString() === model.uri.toString()) {
+                if ((this.context === 'condition') || (this.context === 'logMessage' && DebugBreakpointWidget.isCurlyBracketOpen(inputEditor)) &&
```
Comment (colin-grant-work) [inline packages/debug/src/browser/editor/debug-breakpoint-widget.tsx:96]: It looks like you've changed the bracketing here. Just to be explicit, I'd add parentheses back around the `||` conditions. I might even extract those conditions to named variables, since the condition ends up being pretty long and hard to follow.

File: packages/debug/src/browser/editor/debug-breakpoint-widget.tsx [hunk 2]
```diff
@@ -92,23 +93,30 @@ export class DebugBreakpointWidget implements Disposable {
...
-                        new monaco.suggest.CompletionOptions(undefined, new Set<monaco.languages.CompletionItemKind>().add(monaco.languages.CompletionItemKind.Snippet)),
-                        context, token);
+                    const editorModel = editor.getModel();
+                    const editorPosition = editor.getPosition();
+                    const items = editorModel && editorPosition ? await monaco.suggest.provideSuggestionItems(
+                        editorModel,
+                        new monaco.Position(editorPosition.lineNumber, 1),
+                        new monaco.suggest.CompletionOptions(
+                            undefined,
+                            new Set<monaco.languages.CompletionItemKind>().add(monaco.languages.CompletionItemKind.Snippet),
```
[the remark this answers is not in the document]
Comment (kenneth-marut-work) [reply packages/debug/src/browser/editor/debug-breakpoint-widget.tsx]: 👍

File: packages/debug/src/browser/editor/debug-breakpoint-widget.tsx [hunk 3]
```diff
@@ -84,6 +84,7 @@ export class DebugBreakpointWidget implements Disposable {
         this.zone.containerNode.appendChild(inputNode);
 
         const input = this._input = await this.createInput(inputNode);
+        const inputEditor = input.getControl();
```
[the remark this answers is not in the document]
Comment (kenneth-marut-work) [reply packages/debug/src/browser/editor/debug-breakpoint-widget.tsx]: I've updated the names to be less ambiguous

File: packages/debug/src/browser/editor/debug-breakpoint-widget.tsx (hunk 1 above)
Comment (kenneth-marut-work) [reply packages/debug/src/browser/editor/debug-breakpoint-widget.tsx]: Agreed, I've made some changes based on your suggestions

**Comment by kenneth-marut-work** [comment]
CQ has been submitted, link in the description

File: packages/debug/src/browser/editor/debug-breakpoint-widget.tsx [hunk 4]
```diff
@@ -94,47 +94,63 @@ export class DebugBreakpointWidget implements Disposable {
...
+                    const completionItemModel = (sourceCodeEditorModel && sourceCodeEditorPosition) && await monaco.suggest.provideSuggestionItems(
+                        sourceCodeEditorModel,
+                        new monaco.Position(sourceCodeEditorPosition.lineNumber, 1),
+                        new monaco.suggest.CompletionOptions(
+                            undefined,
+                            new Set<monaco.languages.CompletionItemKind>([monaco.languages.CompletionItemKind.Snippet]),
+                        ),
+                        context,
+                        token,
+                    );
```
Comment (colin-grant-work) [inline packages/debug/src/browser/editor/debug-breakpoint-widget.tsx:123]: It would clean things up a little to move this to its own method. Something like

```typescript
createCompletionItemModel(context, token) {
    const sourceCodeEditor = this.editor.getControl();
    const sourceCodeEditorModel = sourceCodeEditor.getModel();
    const sourceCodeEditorPosition = sourceCodeEditor.getPosition();
    return (sourceCodeEditorModel && sourceCodeEditorPosition) && await monaco.suggest.provideSuggestionItems(
        sourceCodeEditorModel,
        new monaco.Position(sourceCodeEditorPosition.lineNumber, 1),
        new monaco.suggest.CompletionOptions(
            undefined,
            new Set<monaco.languages.CompletionItemKind>([monaco.languages.CompletionItemKind.Snippet]),
        ),
        context,
        token,
    );
}
```

Comment (kenneth-marut-work) [reply packages/debug/src/browser/editor/debug-breakpoint-widget.tsx]: Good idea, I've added this change

**Comment by JonasHelming** [comment]
@colin-grant-work : The CQ has been approved.

**Comment by colin-grant-work** [comment]
@kenneth-marut-work, I believe I've solved the remaining problem with this PR by adding the line `breakpointInputEditor.updateOptions({ suggest: { showWords: false } });` and bringing it up to date with `master` and the Monaco uplift. Would you mind testing it to make sure it does what you intended it to do?

**Review by vince-fugnitto — COMMENTED** [review]
I noticed something which I believe is a bug, but it might be better to confirm.

The first time a `log message` is added we do not get the autocomplate behavior outside of the interpolation which I believe is correct, however the second time we do:


https://user-images.githubusercontent.com/40359487/175310088-9791a2d0-4da9-4ebc-9d10-cf2814e17bb8.mp4

**Comment by colin-grant-work** [comment]
@vince-fugnitto, good catch. I believe what was happening was that the option displayed on the select component was out of synch with the `context` value checked for completions, so it _looked_ like you were entering a log message, but in fact it was being treated as a condition (where completions occur outside of brackets). I've added code to ensure that the context value and the SelectComponent's state are in synch.
